p2p/transport/websocket/conn.go

```diff
@@ -25,7 +28,8 @@ func (c conn) RemoteAddr() net.Addr {
 
 func (c conn) Read(b []byte) (int, error) {
 	n, err := c.Conn.Read(b)
-	if err == nil && n == 0 {
+	if err == nil && n == 0 && c.readAttempts < maxReadAttempts {
```

Comment:
Why is this necessary? An `io.Reader` should not return `n == 0 && err == nil`. Is this a bug upstream?

Reply:
Gorilla did the same thing and we had a for loop then.

From the Go doc on `io.Reader`:

> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0. Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF.

It's discouraged, but allowed to return `0, nil`.

Reply:
Why do we have to deal with it here? Can't we just have the user deal with it?

Reply:
The user is the upgrader in this case, and it doesn't expect this. It fails without this.

Reply:
I'm not sure I understand how this fix helps. If the reader return `0, nil`. I assume that happens when there's no data to read. So retrying immediately will yield the same result, unless new data was received in the few nanoseconds it took to retry the `Read` call.

Really, this sounds like a bug upstream. `Read` should block until new data is available. Otherwise every consumer of this `io.Reader` will into this problem. It's a common pattern to call `Read` in a loop until it errors, and every such loop will be a busy-loop if it returns `0, nil`.

Reply:
>  I assume that happens when there's no data to read. So retrying immediately will yield the same result, unless new data was received in the few nanoseconds it took to retry the Read call.


I think this has to do with the end of a data frame. Before we start a new reader for the next frame: https://github.com/nhooyr/websocket/blob/master/netconn.go#L118. We do the same thing today with Gorilla: https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/websocket/conn.go#L65. The difference here is that we bound it so we aren't looping forever.

Reply:
That makes sense, thank you for the explanation. I'm ok with keeping this workaround for now. What do you think of:
1. Opening an issue upstream. Maybe even submit a PR, this seems pretty straightforward to fix with a few lines of code.
2. Add a comment with your explanation from above (and a link to the issue) to the code.

Reply:
@MarcoPolo Thoughts?

Reply:
done